File: src/types/goal.ts
export type GoalStatus =
  | 'draft'
  | 'active'
  | 'paused'
  | 'blocked'
  | 'completed'
  | 'failed'
  | 'cancelled';

export type GoalPriority = 'low' | 'normal' | 'high' | 'critical';

export type GoalType =
  | 'Project'
  | 'Personal'
  | 'Business'
  | 'Research'
  | 'Learning'
  | 'Maintenance'
  | 'Custom';

export interface Goal {
  id: string;
  worldId: string;
  ownerPersonId: string;
  ownerPersonName?: string;
  ownerPersonEmoji?: string;
  createdBy: string; // 'user' or personId
  title: string;
  description: string;
  type: GoalType;
  status: GoalStatus;
  priority: GoalPriority;
  progress: number; // 0-100 derived from steps
  targetDate?: string;
  activePlanId?: string;
  summary?: string;
  createdAt: string;
  updatedAt: string;
  completedAt?: string;
  metadata?: Record<string, unknown>;
}

File: backend/app/schemas/goal.py
from pydantic import BaseModel
from typing import Optional, List, Any, Dict

class GoalBase(BaseModel):
    worldId: str
    ownerPersonId: str
    ownerPersonName: Optional[str] = None
    ownerPersonEmoji: Optional[str] = None
    createdBy: str = "user"
    title: str
    description: str = ""
    type: str = "Project"
    status: str = "draft"
    priority: str = "normal"
    progress: float = 0
    targetDate: Optional[str] = None
    summary: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = {}

class GoalCreate(GoalBase):
    pass

class GoalUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    type: Optional[str] = None
    status: Optional[str] = None
    priority: Optional[str] = None
    progress: Optional[float] = None
    targetDate: Optional[str] = None
    ownerPersonId: Optional[str] = None
    activePlanId: Optional[str] = None
    summary: Optional[str] = None
    completedAt: Optional[str] = None

class GoalResponse(GoalBase):
    id: str
    activePlanId: Optional[str] = None
    createdAt: str
    updatedAt: str
    completedAt: Optional[str] = None

    class Config:
        from_attributes = True
